### ISSUE TITLE
'process-test' does not add TestCompileSourceRoots to javac '-sourcepath'

### ISSUE THREAD

**Issue by DemonicTutor** [bsorrentino/maven-annotation-plugin#83]
Hi!

Our maven build includes protobuf running before annotation processing for `main` and `test` and therefore the compiler needs to see the compileSourceRoots or otherwise fails with compilation errors because the generated classes from protobuf are missing.

This is working fine for **process** because the path where the generated java files reside is added to `-sourcepath` but for **process-test** it fails because the test-paths are missing.

For **process** it is as expected
`-sourcepath /home/demonictutor/work/workspaces/noenv/microframework/src/main/java:
/home/demonictutor/work/workspaces/noenv/microframework/src/main/generated`

but for **process-test** it is
`-sourcepath /home/demonictutor/work/workspaces/noenv/microframework/src/main/java:
/home/demonictutor/work/workspaces/noenv/microframework/src/main/generated:
/home/demonictutor/work/workspaces/noenv/microframework/target/generated-sources/annotations`

* generated-sources/annotations - don't know why this is set or where its coming from
* src/main/java and src/main/generated - that's fine
* src/test/java and src/test/generated - these are missing

I already have a fix in a branch locally to add the TestCompileSourceRoots which works for our projects at least.

The three relevant plugins from one of our poms:

```
 <plugin>
                    <groupId>org.xolstice.maven.plugins</groupId>
                    <artifactId>protobuf-maven-plugin</artifactId>
                    <version>0.5.1</version>
                    <extensions>true</extensions>
                    <configuration>
                        <protocArtifact>com.google.protobuf:protoc:3.5.1:exe:${os.detected.classifier}</protocArtifact>
                        <pluginId>grpc-java</pluginId>
                        <pluginArtifact>io.vertx:protoc-gen-grpc-java:1.25.0:exe:${os.detected.classifier}</pluginArtifact>
                    </configuration>
                    <executions>
                        <execution>
                            <id>compile</id>
                            <configuration>
                                <outputDirectory>${project.basedir}/${src-main-generated.directory}</outputDirectory>
                                <clearOutputDirectory>false</clearOutputDirectory>
                            </configuration>
                            <goals>
                                <goal>compile</goal>
                                <goal>compile-custom</goal>
                            </goals>
                            <phase>generate-sources</phase>
                        </execution>
                        <execution>
                            <id>test-compile</id>
                            <configuration>
                                <outputDirectory>${project.basedir}/${src-test-generated.directory}</outputDirectory>
                                <clearOutputDirectory>false</clearOutputDirectory>
                            </configuration>
                            <goals>
                                <goal>test-compile</goal>
                                <goal>test-compile-custom</goal>
                            </goals>
                            <phase>generate-test-sources</phase>
                        </execution>
                    </executions>
                </plugin>

                <plugin>
                    <groupId>org.bsc.maven</groupId>
                    <artifactId>maven-processor-plugin</artifactId>
                    <version>4.0</version>
                    <configuration>
                        <addOutputDirectoryToCompilationSources>true</addOutputDirectoryToCompilationSources>
                        <compilerArguments>-implicit:none</compilerArguments>
                        <systemProperties>
                            <java.util.logging.SimpleFormatter.format>%4$s: %3$s - %5$s %6$s%n</java.util.logging.SimpleFormatter.format>
                        </systemProperties>
                        <processors>
                            <processor>io.vertx.codegen.CodeGenProcessor</processor>
                        </processors>
                    </configuration>
                    <executions>
                        <execution>
                            <id>generate-sources</id>
                            <goals>
                                <goal>process</goal>
                            </goals>
                            <phase>generate-sources</phase>
                            <configuration>
                                <optionMap>
                                  <codegen.output>${project.basedir}/src/main</codegen.output>
                                  <codegen.output.data_object_converters>generated</codegen.output.data_object_converters>
                                  <codegen.output.cheatsheet>asciidoc</codegen.output.cheatsheet>
                                </optionMap>
                                <defaultOutputDirectory>${project.basedir}/${src-main-generated.directory}</defaultOutputDirectory>
                            </configuration>
                        </execution>
                        <execution>
                            <id>generate-test-sources</id>
                            <goals>
                                <goal>process-test</goal>
                            </goals>
                            <phase>generate-test-sources</phase>
                            <configuration>
                                <optionMap>
                                    <codegen.output>${project.basedir}/src/test</codegen.output>
                                    <codegen.output.data_object_converters>generated</codegen.output.data_object_converters>
                                    <codegen.output.cheatsheet>asciidoc</codegen.output.cheatsheet>
                                </optionMap>
                                <defaultOutputDirectory>${project.basedir}/${src-test-generated.directory}</defaultOutputDirectory>
                            </configuration>
                        </execution>
                    </executions>
                </plugin>

                <plugin>
                    <artifactId>maven-compiler-plugin</artifactId>
                    <version>3.8.1</version>
                    <configuration>
                        <source>${maven.compiler.source}</source>
                        <target>${maven.compiler.target}</target>
                        <compilerArgs>
                            <arg>-Xlint:unchecked</arg>
                            <arg>-Xlint:deprecation</arg>
                            <arg>-Xlint:removal</arg>
                        </compilerArgs>
                    </configuration>
                </plugin>
```

output from `mvn clean package -DskipTests -X` 

```
Configuring mojo 'org.bsc.maven:maven-processor-plugin:4.0:process' with basic configurator -->
  (f) addCompileSourceRoots = false
  (f) addOutputDirectoryToCompilationSources = true
  (f) appendSourceArtifacts = false
  (f) classpathElements = [/home/demonictutor/work/workspaces/noenv/microframework/target/classes, <<< CUT >>>]
  (f) compilerArguments = -implicit:none
  (f) defaultOutputDirectory = /home/demonictutor/work/workspaces/noenv/microframework/src/main/generated
  (f) encoding = UTF-8
  (f) failOnError = true
  (f) fork = false
  (f) optionMap = {codegen.output=/home/demonictutor/work/workspaces/noenv/microframework/src/main, codegen.output.cheatsheet=asciidoc, codegen.output.data_object_converters=generated}
  (f) outputClassDirectory = /home/demonictutor/work/workspaces/noenv/microframework/target/classes
  (f) outputDiagnostics = true
  (f) pluginArtifacts = [org.bsc.maven:maven-processor-plugin:maven-plugin:4.0: <<< CUT >>>]
  (f) processors = [io.vertx.codegen.CodeGenProcessor]
  (f) project = MavenProject: com.noenv:microframework-core:1.5-SNAPSHOT @ /home/demonictutor/work/workspaces/noenv/microframework/pom.xml
  (f) remoteRepos = [nexus-noenv-releases (https://nexus.noenv.com/repository/maven-releases/, default, releases), nexus-noenv-snapshots (https://nexus.noenv.com/repository/maven-snapshots/, default, releases+snapshots), central (https://repo.maven.apache.org/maven2, default, releases)]
  (f) repoSession = org.eclipse.aether.DefaultRepositorySystemSession@63f9b562
  (f) session = org.apache.maven.execution.MavenSession@6993c8df
  (f) skip = false
  (f) sourceDirectory = /home/demonictutor/work/workspaces/noenv/microframework/src/main/java
  (f) systemProperties = {java.util.logging.SimpleFormatter.format=%4$s: %3$s - %5$s %6$s%n}
-- end configuration --
Source directory: /home/demonictutor/work/workspaces/noenv/microframework/src/main/generated added
processing source directory [/home/demonictutor/work/workspaces/noenv/microframework/src/main/java]
Adding compiler arg: -implicit:none
Adding compiler arg: -Acodegen.output=/home/demonictutor/work/workspaces/noenv/microframework/src/main
Adding compiler arg: -Acodegen.output.cheatsheet=asciidoc
Adding compiler arg: -Acodegen.output.data_object_converters=generated
javac option: -cp
javac option: /home/demonictutor/work/workspaces/noenv/microframework/src/main/resources:<<< CUT >>>
javac option: --module-path
javac option: /home/demonictutor/work/workspaces/noenv/microframework/target/classes: <<< CUT >>>
javac option: -sourcepath
javac option: /home/demonictutor/work/workspaces/noenv/microframework/src/main/java:/home/demonictutor/work/workspaces/noenv/microframework/src/main/generated
javac option: -proc:only
javac option: -implicit:none
javac option: -Acodegen.output=/home/demonictutor/work/workspaces/noenv/microframework/src/main
javac option: -Acodegen.output.cheatsheet=asciidoc
javac option: -Acodegen.output.data_object_converters=generated
javac option: -processor
javac option: io.vertx.codegen.CodeGenProcessor
javac option: -d
javac option: /home/demonictutor/work/workspaces/noenv/microframework/target/classes
javac option: -s
javac option: /home/demonictutor/work/workspaces/noenv/microframework/src/main/generated
javac option: -source
javac option: 14
javac option: -target
javac option: 14
set system property : [java.util.logging.SimpleFormatter.format] = [%4$s: %3$s - %5$s %6$s%n]
```

```
 Configuring mojo org.bsc.maven:maven-processor-plugin:4.0:process-test from plugin realm ClassRealm[plugin>org.bsc.maven:maven-processor-plugin:4.0, parent: jdk.internal.loader.ClassLoaders$AppClassLoader@5bc2b487]
 Configuring mojo 'org.bsc.maven:maven-processor-plugin:4.0:process-test' with basic configurator -->
   (f) addCompileSourceRoots = true
   (f) addOutputDirectoryToCompilationSources = true
   (f) appendSourceArtifacts = false
   (f) classpathElements = [/home/demonictutor/work/workspaces/noenv/microframework/target/test-classes, /home/demonictutor/work/workspaces/noenv/microframework/target/classes, <<< CUT >>>]
   (f) compilerArguments = -implicit:none
   (f) defaultOutputDirectory = /home/demonictutor/work/workspaces/noenv/microframework/src/test/generated
   (f) encoding = UTF-8
   (f) failOnError = true
   (f) fork = false
   (f) optionMap = {codegen.output=/home/demonictutor/work/workspaces/noenv/microframework/src/test, codegen.output.cheatsheet=asciidoc, codegen.output.data_object_converters=generated}
   (f) outputClassDirectory = /home/demonictutor/work/workspaces/noenv/microframework/target/test-classes
   (f) outputDiagnostics = true
   (f) pluginArtifacts = [org.bsc.maven:maven-processor-plugin:maven-plugin:4.0:<<< CUT >>>]
   (f) processors = [io.vertx.codegen.CodeGenProcessor]
   (f) project = MavenProject: com.noenv:microframework-core:1.5-SNAPSHOT @ /home/demonictutor/work/workspaces/noenv/microframework/pom.xml
   (f) remoteRepos = [nexus-noenv-releases (https://nexus.noenv.com/repository/maven-releases/, default, releases), nexus-noenv-snapshots (https://nexus.noenv.com/repository/maven-snapshots/, default, releases+snapshots), central (https://repo.maven.apache.org/maven2, default, releases)]
   (f) repoSession = org.eclipse.aether.DefaultRepositorySystemSession@63f9b562
   (f) session = org.apache.maven.execution.MavenSession@6993c8df
   (f) skip = false
   (f) sourceDirectory = /home/demonictutor/work/workspaces/noenv/microframework/src/test/java
   (f) systemProperties = {java.util.logging.SimpleFormatter.format=%4$s: %3$s - %5$s %6$s%n}
 -- end configuration --
 Source directory: /home/demonictutor/work/workspaces/noenv/microframework/src/test/generated added
 processing source directory [/home/demonictutor/work/workspaces/noenv/microframework/src/test/java]
 processing source directory [/home/demonictutor/work/workspaces/noenv/microframework/src/main/java]
 processing source directory [/home/demonictutor/work/workspaces/noenv/microframework/target/generated-sources/annotations]
 processing source directory [/home/demonictutor/work/workspaces/noenv/microframework/src/main/generated]
 Adding compiler arg: -implicit:none
 Adding compiler arg: -Acodegen.output=/home/demonictutor/work/workspaces/noenv/microframework/src/test
 Adding compiler arg: -Acodegen.output.cheatsheet=asciidoc
 Adding compiler arg: -Acodegen.output.data_object_converters=generated
 javac option: -cp
 javac option: /home/demonictutor/work/workspaces/noenv/microframework/src/test/resources:/home/demonictutor/work/workspaces/noenv/microframework/src/test/proto:/home/demonictutor/work/workspaces/noenv/microframework/target/test-classes:/home/demonictutor/work/workspaces/noenv/microframework/target/classes<<< CUT >>>
 javac option: --module-path
 javac option: /home/demonictutor/work/workspaces/noenv/microframework/target/test-classes:/home/demonictutor/work/workspaces/noenv/microframework/target/classes:<<< CUT >>>
 javac option: -sourcepath
 javac option: /home/demonictutor/work/workspaces/noenv/microframework/src/main/java:/home/demonictutor/work/workspaces/noenv/microframework/src/main/generated:/home/demonictutor/work/workspaces/noenv/microframework/target/generated-sources/annotations
 javac option: -proc:only
 javac option: -implicit:none
 javac option: -Acodegen.output=/home/demonictutor/work/workspaces/noenv/microframework/src/test
 javac option: -Acodegen.output.cheatsheet=asciidoc
 javac option: -Acodegen.output.data_object_converters=generated
 javac option: -processor
 javac option: io.vertx.codegen.CodeGenProcessor
 javac option: -d
 javac option: /home/demonictutor/work/workspaces/noenv/microframework/target/test-classes
 javac option: -s
 javac option: /home/demonictutor/work/workspaces/noenv/microframework/src/test/generated
 javac option: -source
 javac option: 14
 javac option: -target
 javac option: 14
 set system property : [java.util.logging.SimpleFormatter.format] = [%4$s: %3$s - %5$s %6$s%n]
```

**Comment by bsorrentino** [comment]
Hi, @DemonicTutor  thanks for the detailed feedback.

I'll be starting work on it as soon as possible. However, when you are able to fix the code, always feel free to make a PR

**Comment by DemonicTutor** [comment]
guess i failed at linking it properly.. but its there :/

**Comment by bsorrentino** [comment]
Hi @DemonicTutor 

I've merged #84 and delivered the developer version `4.2-SNAPSHOT`

Do you have a chance to test it and let me know?

Thanks in advance

**Comment by DemonicTutor** [comment]
Hey @bsorrentino !

Thanks - my projects are building so it seems to be fine :+1:
